x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/SourceConfig.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.featureindexbuilder.job;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeValuesSourceBuilder;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeValuesSourceParserHelper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
+/*
+ * Wrapper for the Source config part of a composite aggregation.
+ *
+ * For now just wraps sources from composite aggs.
+ */
+public class SourceConfig implements Writeable, ToXContentObject {
+
+    private static final String NAME = "feature_index_builder_source";
+
+    private final List<CompositeValuesSourceBuilder<?>> sources;
+
+    private static final ConstructingObjectParser<SourceConfig, String> PARSER = new ConstructingObjectParser<>(NAME, false, (args) -> {
+        @SuppressWarnings("unchecked")
+        List<CompositeValuesSourceBuilder<?>> sources = (List<CompositeValuesSourceBuilder<?>>) args[0];
+        return new SourceConfig(sources);
+    });
+
+    static {
+        PARSER.declareFieldArray(constructorArg(), (parser, builder) -> CompositeValuesSourceParserHelper.fromXContent(parser),
+                CompositeAggregationBuilder.SOURCES_FIELD_NAME, ObjectParser.ValueType.OBJECT_ARRAY);
+    }
+
+    SourceConfig(final StreamInput in) throws IOException {
+        int num = in.readVInt();
+        this.sources = new ArrayList<>(num);
+        for (int i = 0; i < num; i++) {
+            CompositeValuesSourceBuilder<?> builder = CompositeValuesSourceParserHelper.readFrom(in);
+            getSources().add(builder);
```

Comment:
It would be nice if this class was immutable, and this line shows why it isn't currently.  This is how to make it immutable:

```
        List<CompositeValuesSourceBuilder<?>> sources = new ArrayList<>(num);
        for (int i = 0; i < num; i++) {
            CompositeValuesSourceBuilder<?> builder = CompositeValuesSourceParserHelper.readFrom(in);
            sources.add(builder);
        }
        this.sources = Collections.unmodifableList(sources);
```